internal/datasource/insights.go

```diff
@@ -0,0 +1,95 @@
+package datasource
+
+import (
+	"context"
+	"crypto/x509"
+	"fmt"
+	"sync"
+	"time"
+
+	pb "deps.dev/api/v3"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials"
+)
+
+// CachedInsightsClient is a wrapper for InsightsClient that caches requests.
+type CachedInsightsClient struct {
+	pb.InsightsClient
+
+	// cache fields
+	mu                sync.Mutex
+	cacheTimestamp    *time.Time
+	packageCache      *RequestCache[packageKey, *pb.Package]
+	versionCache      *RequestCache[versionKey, *pb.Version]
+	requirementsCache *RequestCache[versionKey, *pb.Requirements]
+}
+
+// Comparable types to use as map keys for cache.
+type packageKey struct {
+	System pb.System
+	Name   string
+}
+
+func makePackageKey(k *pb.PackageKey) packageKey {
+	return packageKey{
+		System: k.GetSystem(),
+		Name:   k.GetName(),
+	}
+}
+
+type versionKey struct {
+	System  pb.System
+	Name    string
+	Version string
+}
+
+func makeVersionKey(k *pb.VersionKey) versionKey {
+	return versionKey{
+		System:  k.GetSystem(),
+		Name:    k.GetName(),
+		Version: k.GetVersion(),
+	}
+}
+
+func NewCachedInsightsClient(addr string, userAgent string) (*CachedInsightsClient, error) {
```

Comment:
Internal linters require all public functions to have a comment starting with the function name
e.g. // NewCachedInsightClient creates a CachedInsightsClient